[Bug]: PR 17365 breaks BQ loads in some cases

### What happened?

In beam 2.39, this used to work:
```
BigQueryIO
   .write()
   .optimizedWrites()
   .to(OutputTable)
   .withAvroFormatFunction(abc)
   .withAvroSchemaFactory(xyz)
   .withCreateDisposition(CreateDisposition.CREATE_NEVER)
   .withWriteDisposition(WriteDisposition.WRITE_TRUNCATE))
```

but now throws an exception if you end up in MultiPartitionsWriteTables:

```
Caused by: java.lang.IllegalArgumentException: Unless create disposition is CREATE_NEVER, a schema must be specified, i.e. DynamicDestinations.getSchema() may not return null. However, create disposition is CREATE_IF_NEEDED, and org.apache.beam.sdk.io.gcp.bigquery.DynamicDestinationsHelpers$ConstantTableDestinations@52e95e1 returned null for destination tableSpec: <table>
	at org.apache.beam.sdk.util.Preconditions.checkArgumentNotNull(Preconditions.java:436)
	at org.apache.beam.sdk.io.gcp.bigquery.WriteTables$WriteTablesDoFn.processElement(WriteTables.java:207)
```

The reason seems like now, the create disposition is set to `CreateDisposition.CREATE_IF_NEEDED` in WriteTables no matter what the user sets it to.

cc @pabloem @MarcoRob 

### Issue Priority

Priority: 2

### Issue Component

Component: io-java-gcp

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
@@ -719,9 +719,12 @@ class BatchLoads<DestinationT, ElementT>
     // If the final destination table exists already (and we're appending to it), then the temp
     // tables must exactly match schema, partitioning, etc. Wrap the DynamicDestinations object
     // with one that makes this happen.
-    DynamicDestinations<?, DestinationT> matchedDestinations =
-        DynamicDestinationsHelpers.matchTableDynamicDestinations(
-            dynamicDestinations, bigQueryServices);
+    // In the case schemaUpdateOptions are specified by the user, matching does not occur in order
+    // to respect those options.
+    DynamicDestinations<?, DestinationT> destinations = dynamicDestinations;
+    if (schemaUpdateOptions.isEmpty()) {
+      destinations = DynamicDestinationsHelpers.matchTableDynamicDestinations(dynamicDestinations, bigQueryServices);
+    }
 
     // If WriteBundlesToFiles produced more than DEFAULT_MAX_FILES_PER_PARTITION files or
     // DEFAULT_MAX_BYTES_PER_PARTITION bytes, then
@@ -741,7 +744,7 @@ class BatchLoads<DestinationT, ElementT>
                 WriteDisposition.WRITE_EMPTY,
                 CreateDisposition.CREATE_IF_NEEDED,
                 sideInputs,
-                matchedDestinations,
+                destinations,
                 loadJobProjectId,
                 maxRetryJobs,
                 ignoreUnknownValues,

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
@@ -723,7 +723,9 @@ class BatchLoads<DestinationT, ElementT>
     // to respect those options.
     DynamicDestinations<?, DestinationT> destinations = dynamicDestinations;
     if (schemaUpdateOptions.isEmpty()) {
-      destinations = DynamicDestinationsHelpers.matchTableDynamicDestinations(dynamicDestinations, bigQueryServices);
+      destinations =
+          DynamicDestinationsHelpers.matchTableDynamicDestinations(
+              dynamicDestinations, bigQueryServices);
     }
 
     // If WriteBundlesToFiles produced more than DEFAULT_MAX_FILES_PER_PARTITION files or

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
@@ -716,6 +716,13 @@ class BatchLoads<DestinationT, ElementT>
             ShardedKeyCoder.of(NullableCoder.of(destinationCoder)),
             WritePartition.ResultCoder.INSTANCE);
 
+    // If the final destination table exists already (and we're appending to it), then the temp
+    // tables must exactly match schema, partitioning, etc. Wrap the DynamicDestinations object
+    // with one that makes this happen.
+    DynamicDestinations<?, DestinationT> matchedDestinations =
+        DynamicDestinationsHelpers.matchTableDynamicDestinations(
+            dynamicDestinations, bigQueryServices);
+
     // If WriteBundlesToFiles produced more than DEFAULT_MAX_FILES_PER_PARTITION files or
     // DEFAULT_MAX_BYTES_PER_PARTITION bytes, then
     // the import needs to be split into multiple partitions, and those partitions will be
@@ -734,7 +741,7 @@ class BatchLoads<DestinationT, ElementT>
                 WriteDisposition.WRITE_EMPTY,
                 CreateDisposition.CREATE_IF_NEEDED,
                 sideInputs,
-                dynamicDestinations,
+                matchedDestinations,
                 loadJobProjectId,
                 maxRetryJobs,
                 ignoreUnknownValues,

